dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -1328,6 +1329,86 @@ private List<Event> inlineMemCpy(FunctionCall call) {
+        final Expression src = call.getArguments().get(2);
+        final Expression countExpr = call.getArguments().get(3);
+
+        if (!(countExpr instanceof IntLiteral countValue)) {
+            final String error = "Cannot handle memcpy_s with dynamic count argument: " + call;
+            throw new UnsupportedOperationException(error);
+        }
+        final int count = countValue.getValueAsInt();
+        if (!(destszExpr instanceof IntLiteral destszValue)) {
+            final String error = "Cannot handle memcpy_s with dynamic destsz argument: " + call;
+            throw new UnsupportedOperationException(error);
+        }
+        final int destsz = destszValue.getValueAsInt();
+
+        // Runtime checks
+        final Expression nullExpr = expressions.makeZero(types.getArchType());
+        final Expression destIsNull = expressions.makeEQ(dest, nullExpr);
+        final Expression srcIsNull = expressions.makeEQ(src, nullExpr);
+        // We assume RSIZE_MAX = 2^64-1
+        final Expression rsize_max = expressions.makeValue(BigInteger.ONE.shiftLeft(64).subtract(BigInteger.ONE), types.getArchType());
+        final Expression gtMax = expressions.makeOr(
+                expressions.makeGT(countExpr, rsize_max, false),
+                expressions.makeGT(destszExpr, rsize_max, false));
+        final Expression countGtdestszExpr = expressions.makeGT(countExpr, destszExpr, false);
+        final Expression overlap = expressions.makeOr(
+                expressions.makeGTE(expressions.makeAdd(src, countExpr), dest, false),
+                expressions.makeGTE(expressions.makeAdd(dest, countExpr), src, false));
+
+        // Runtime full condition
+        final Expression c1 = expressions.makeOr(destIsNull, srcIsNull);
+        final Expression c2 = expressions.makeOr(c1, gtMax);
+        final Expression c3 = expressions.makeOr(c2, countGtdestszExpr);
+        final Expression c4 = expressions.makeOr(c3, overlap);
+
+        final List<Event> replacement = new ArrayList<>(2 * count + 1);
+        for (int i = 0; i < count; i++) {
+            final Expression offset = expressions.makeValue(i, types.getArchType());
+            final Expression srcAddr = expressions.makeAdd(src, offset);
+            final Expression destAddr = expressions.makeAdd(dest, offset);
+            // FIXME: We have no other choice but to load ptr-sized chunks for now
+            final Register reg = caller.getOrNewRegister("__memcpy_s_" + i, types.getArchType());
+            // We should zero out the entire destination range [dest, dest+destsz)
+            // Here we handle [dest, dest+count), the remaining are handled by the loop below
+            final Expression value = expressions.makeITE(c4, expressions.makeZero(types.getArchType()), reg);
+
+            replacement.addAll(List.of(
+                    EventFactory.newLoad(reg, srcAddr),
+                    EventFactory.newStore(destAddr, value)
+            ));
```

Comment:
There are three cases for `memcpy_s`:
- The inputs are valid and you do a standard memcpy (+ generate a success error code).
- The inputs are invalid and in particular `dest == NULL`, then the method only produces an error code without any memory accesses.
- The inputs are invalid but at least `dest != NULL` (*), then the range `[dest, dest + destsz]` gets zeroed which only requires memory stores (no loads).

I don't think your implementation is correct. You have to use proper branching to make the case distinctions rather than trying to use ITE.

Btw. how is this function even used in practice? Does the compiler magically transform `memcpy` to `memcpy_s` or does the user specficially call the latter? If so, to what extent can the user recover from an error or does he just terminate the program but with an error code that is better than segfault?
Basically, I'm wondering if failure of `memcpy_s` is a bug in pretty much all cases. If so, we could also generate an `assert (boundCheck) ### memcpy_s error` and don't even bother too much about the failure details (we could also provide an option to toggle this behaviour).

(*) Technically, `destsz` could also be invalid but only if it is larger than `RSIZE_MAX` which is impossible in the implementation anyway.

Reply:
Ok, I missed the "(if both dest and destsz are valid)". After re-reading, this is my understanding
- `dest == NULL` or `invalid destsz` ----> return error > 0
- `dest != NULL && valid destsz && (src == NULL || count > destsz || overlap(src, dest))` ----> return error > 0 and zero out `[dest, dest+destsz)`

> I don't think your implementation is correct. You have to use proper branching to make the case distinctions rather than trying to use ITE.

Why do you think we need branches? I don't see this.

> Btw. how is this function even used in practice? Does the compiler magically transform `memcpy` to `memcpy_s` or does the user specficially call the latter? 

The user explicitly uses the later. This is used in libvsync and the whole reason of this PR. It is then the responsibility of the developer to check if the returned value != 0.

> If so, to what extent can the user recover from an error or does he just terminate the program but with an error code that is better than segfault? Basically, I'm wondering if failure of `memcpy_s` is a bug in pretty much all cases. If so, we could also generate an `assert (boundCheck) ### memcpy_s error` and don't even bother too much about the failure details (we could also provide an option to toggle this behaviour).

This is up to the user. If we want to be precise about its semantics, I don't think we can make this.
If we don't want to be precise, I would simply treat `memcpy_s` as `memcpy` and get minimal changes to the current code.

> 
> (*) Technically, `destsz` could also be invalid but only if it is larger than `RSIZE_MAX` which is impossible in the implementation anyway.

Reply:
> Ok, I missed the "(if both dest and destsz are valid)". After re-reading, this is my understanding
> 
>     * `dest == NULL` or `invalid destsz` ----> return error > 0
> 
>     * `dest != NULL && valid destsz && (src == NULL || count > destsz || overlap(src, dest))` ----> return error > 0 and zero out `[dest, dest+destsz)`

Yes, that seems right. Importantly, in neither of those two cases you load from `src`!

> 
> 
> > I don't think your implementation is correct. You have to use proper branching to make the case distinctions rather than trying to use ITE.
> 
> Why do you think we need branches? I don't see this.

You just listed the different cases. One case performs loads and stores, another only stores, and the third one has no memory accesses. There is no way to get this behaviour without three different control-flow branches.
Your solution tried to merge cases (1) and (2) by always performing the loads and just ignoring their results to mimic case (2). These redundant loads are already questionable and you should avoid them, but you definitely cannot have redundant stores so case (3) must be different than the other two.